chanfitness/chanfitness.go

```diff
@@ -0,0 +1,300 @@
+// Package chanfitness monitors the behaviour of channels to provide scores
+// which can be factored into the decision to close a channel. This is
+// achieved by maintaining an event store which tracks events for each
+// channel.
+//
+// Lifetime: the total time the channel has been known to the scoring system.
+// Note that this does not equal channel lifetime because data is not currently
+// persisted.
+//
+// Uptime: the total time within the channel's lifetime that its peer has been
+// observed as being online.
+package chanfitness
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/peernotifier"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ChannelEventStore maintains a set of event logs for the node's channels.
+// It is intended to provide insight into the performance of channels.
+type ChannelEventStore struct {
+	cfg *Config
+
+	// channels maps short channel IDs to event logs.
+	channels map[uint64]*chanEventLog
+
+	quit chan struct{}
+
+	sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
+
+	// PeerOnline return true if a peer is online, and false if it is offline
+	// or unknown to the server.
+	PeerOnline func(peer [33]byte) bool
```

Comment:
If every peer is assumed offline on startup and the peers would all push an event into the channel when they go online on startup, is this second pathway still necessary? Tracking started already when the open channels were listed?

Reply:
This is necessary for new channels that are opened after startup. We don't know whether the peer is online when the channel confirms, and would have ignored the most recent peer event before we had a channel open with them so need some way of figuring out their starting state. We could technically infer a state if they have a single online/offline event while we're monitoring, but if they're online or offline for our whole monitoring period we don't know which state they're in.

Reply:
If you want to keep the interface with the fitness system as narrow as possible (which I think is good), you could just have an in-memory map on the event store level that stores the peers that are online. Do the look up on that locally inside the fitness system. If the peer is not in the map, it is offline. If the peer is online for the whole monitoring period, the peer would still have send an online event on startup to the fitness system?

Reply:
Ok will do. I was thinking that current peer state (and perhaps a history of online/offline events to get around potential duplication from per channel tracking with the same peer) could be added to `PeerNotifier` if there's demand for something like that in more than one place. But keeping it in `chanfitness` in the meantime works.